libs/Microsoft.MixedReality.WebRTC.Unity/Assets/Microsoft.MixedReality.WebRTC.Unity.Editor/WebcamSourceEditor.cs

```diff
@@ -0,0 +1,525 @@
+                                EditorGUILayout.HelpBox("The SDP codec name must be non-empty. See https://en.wikipedia.org/wiki/RTP_audio_video_profile for valid names.", MessageType.Error);
+
+                                // Force a non-empty value now, otherwise the field will reset to None
+                                newValue = "__CUSTOM";
+                            }
+                            _preferredVideoCodec.stringValue = newValue;
+                            --EditorGUI.indentLevel;
+                        }
+                        else
+                        {
+                            _preferredVideoCodec.stringValue = Enum.GetName(typeof(SdpVideoCodecs), newCodecValue);
+                        }
+                    }
+                }
+                catch (Exception)
+                {
+                    EditorGUILayout.PropertyField(_preferredVideoCodec);
+                }
+            }
+
+            serializedObject.ApplyModifiedProperties();
+        }
+
+        /// <summary>
+        /// ToggleLeft control associated with a given SerializedProperty, to enable automatic GUI
+        /// handlings like Prefab revert menu.
+        /// </summary>
+        /// <param name="property">The boolean property associated with the control.</param>
+        /// <param name="label">The label to display next to the toggle control.</param>
+        private void ToggleLeft(SerializedProperty property, GUIContent label)
+        {
+            var rect = EditorGUILayout.GetControlRect();
+            using (new EditorGUI.PropertyScope(rect, label, property))
+            {
+                property.boolValue = EditorGUI.ToggleLeft(rect, label, property.boolValue);
+            }
+        }
+
+        /// <summary>
+        /// IntField with optional toggle associated with a given SerializedProperty, to enable
+        /// automatic GUI handlings like Prefab revert menu.
+        /// </summary>
+        /// <param name="intProperty">The integer property associated with the control.</param>
+        /// <param name="defaultValue">Default value if the property value is negative or zero. Assigned the new value on return if valid.</param>
+        /// <param name="label">The label to display next to the toggle control.</param>
+        /// <param name="unitLabel">The label indicating the unit of the value.</param>
+        private void OptionalIntField(SerializedProperty intProperty, ref int defaultValue, GUIContent label, GUIContent unitLabel)
+        {
+            if (defaultValue <= 0)
+            {
+                throw new ArgumentOutOfRangeException("Default value cannot be invalid.");
+            }
+
+            using (new EditorGUILayout.HorizontalScope())
+            {
+                var rect = EditorGUILayout.GetControlRect();
+                using (new EditorGUI.PropertyScope(rect, label, intProperty))
+                {
+                    bool hasConstraint = (intProperty.intValue > 0);
```

Comment:
All this constraint stuff is a bit obscure. The name `OptionalIntField` makes me think to something like `std::optional<int>` so I can't make much sense of why constraints and explicit comparisons to 0 are involved. Should be documented.

Reply:
The problem is you have to work with Unity serializing only the `int` value. So you cannot add a `bool` to indicate the optional. The way this works in the interop API is that zero/negative values are considered to be invalid, and mean that the field does not have a value. Here we reuse that same pattern. The word "constraint" here is a leftover; I can reword. But the general mechanism stays, I don't have any alternative for it.

Reply:
Still a few issues:
- `defaultValue` is not a great name since it's not really a default. `lastValidValue` maybe?
- `hasValidValue` is used twice with the meaning of "property is set to a valid value" first and "checkbox is set" aka "must get a new value with a text field" later. This is quite confusing, should be two variables. 
- there are a few instances of "constraint" left over

Reply:
Good suggestions, fixed all.